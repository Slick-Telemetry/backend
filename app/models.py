# Built-in
from typing import Dict, List

# External
from pydantic import BaseModel, RootModel


class Root(BaseModel):
    """Response model for root"""

    we_are: str = "SlickTelemetry"


class EventSchedule(BaseModel):
    """Response model for event schedule data for a Formula 1 calendar year"""

    RoundNumber: int
    Country: str
    Location: str
    OfficialEventName: str
    EventDate: str
    EventName: str
    EventFormat: str
    Session1: str
    Session1Date: str
    Session1DateUtc: str
    Session2: str
    Session2Date: str
    Session2DateUtc: str
    Session3: str
    Session3Date: str
    Session3DateUtc: str
    Session4: str
    Session4Date: str
    Session4DateUtc: str
    Session5: str
    Session5Date: str
    Session5DateUtc: str
    F1ApiSupport: bool


class Schedule(BaseModel):
    """Response model for event schedule data with year"""

    year: int
    EventSchedule: List[EventSchedule]


class HealthCheck(BaseModel):
    """Response model to validate and return when performing a health check"""

    status: str = "OK"


class Driver(BaseModel):
    """Model for driver data"""

    driverId: str
    permanentNumber: str
    code: str
    url: str
    givenName: str
    familyName: str
    dateOfBirth: str
    nationality: str


class Constructor(BaseModel):
    """Model for constructor data"""

    constructorId: str
    url: str
    name: str
    nationality: str


class DriverStandings(BaseModel):
    """Model for driver standings data"""

    position: str
    positionText: str
    points: str
    wins: str
    Driver: Driver
    Constructors: List[Constructor]


class ConstructorStandings(BaseModel):
    """Model for constructor standings data"""

    position: str
    positionText: str
    points: str
    wins: str
    Constructor: Constructor


class Standings(BaseModel):
    """Response model for driver and constructor standings for a given season and round"""

    season: int
    round: int
    DriverStandings: List[DriverStandings]
    ConstructorStandings: List[ConstructorStandings]


class Results(BaseModel):
    """Response model for session results for a given year, round and session"""

    DriverNumber: str | None
    BroadcastName: str | None
    Abbreviation: str | None
    DriverId: str | None
    TeamName: str | None
    TeamColor: str | None
    TeamId: str | None
    FirstName: str | None
    LastName: str | None
    FullName: str | None
    HeadshotUrl: str | None
    CountryCode: str | None
    Position: float | None
    ClassifiedPosition: str | None
    GridPosition: float | None
    Q1: int | None
    Q2: int | None
    Q3: int | None
    Time: int | None
    Status: str | None
    Points: float | None


class Laps(BaseModel):
    """Response model for session laps for a given year, round, session and drivers"""

    Time: int | None
    Driver: str | None
    DriverNumber: str | None
    LapTime: int | None
    LapNumber: float | None
    Stint: float | None
    PitOutTime: int | None
    PitInTime: int | None
    Sector1Time: int | None
    Sector2Time: int | None
    Sector3Time: int | None
    Sector1SessionTime: int | None
    Sector2SessionTime: int | None
    Sector3SessionTime: int | None
    SpeedI1: float | None
    SpeedI2: float | None
    SpeedFL: float | None
    SpeedST: float | None
    IsPersonalBest: bool | None
    Compound: str | None
    TyreLife: float | None
    FreshTyre: bool | None
    Team: str | None
    LapStartTime: int | None
    LapStartDate: str | None
    TrackStatus: str | None
    Position: float | None
    Deleted: bool | None
    DeletedReason: str | None
    FastF1Generated: bool | None
    IsAccurate: bool | None


class SplitQualifyingLaps(BaseModel):
    """Response model for split qualifying laps"""

    Q1: List[Laps] | None
    Q2: List[Laps] | None
    Q3: List[Laps] | None


class Telemetry(BaseModel):
    """Response model for session telemetry for a given year, round, session, driver and laps"""

    Time: int
    RPM: int
    Speed: float
    nGear: int
    Throttle: float
    Brake: bool
    DRS: int
    Distance: float
    X: float
    Y: float


class Weather(BaseModel):
    """Response model for session weather for a given year, round, session and laps"""

    Time: int
    AirTemp: float
    Humidity: float
    Pressure: float
    Rainfall: bool
    TrackTemp: float
    WindDirection: int
    WindSpeed: float


class ExtendedTelemetry(BaseModel):
    """Response model for telemetry with weather"""

    Telemetry: List[Telemetry]
    Weather: Weather | None


class DriverSectorTimes(BaseModel):
    MinSector1Time: float
    MinSector2Time: float
    MinSector3Time: float


class FastestSectors(RootModel):
    """Response model for fastest sectors for a given year, round and session"""

    Dict[str, DriverSectorTimes]
