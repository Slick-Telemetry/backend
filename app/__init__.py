import newrelic.agent  # isort:skip


# Built-in
import os


if os.getenv("ENVIRONMENT") != "TEST":
    newrelic.agent.initialize("newrelic.ini")

# External
import fastf1
from dotenv import dotenv_values
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from fastapi.security import HTTPBearer
from fastf1.ergast import Ergast


__version__ = "0.6.0"


# Load environment variables from .env file
config = dotenv_values(".env")
# FastF1 configuration
fastf1.set_log_level("WARNING")
fastf1.Cache.set_disabled()
# Ergast configuration
ergast = Ergast(result_type="raw", auto_cast=True)
# Cors Middleware
origins = [config["FRONTEND_URL"], "http://127.0.0.1:3000"]
# Others
favicon_path = "favicon.ico"
# Security
security = HTTPBearer()

app = FastAPI(
    title="Slick Telemetry API",
    description="Slick Telemetry backend written in python with fastf1. üèé",
    version=__version__,
    contact={
        "name": "Slick Telemetry",
        "url": "https://github.com/Slick-Telemetry",
        "email": "contact@slicktelemetry.com",
    },
    license_info={
        "name": "GNU General Public License v3.0",
        "url": "https://www.gnu.org/licenses/gpl-3.0.en.html",
    },
    redoc_url=None,
)
app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
    # HTTPSRedirectMiddleware # TODO use for production and staging
)
