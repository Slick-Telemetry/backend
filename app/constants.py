# Built-in
from datetime import datetime


METADATA_DESCRIPTION = """
Slick Telemetry backend written in python with fastf1. üèé 
"""

EVENT_SCHEDULE_DATETIME_DTYPE_LIST = [
    "EventDate",
    "Session1Date",
    "Session1DateUtc",
    "Session2Date",
    "Session2DateUtc",
    "Session3Date",
    "Session3DateUtc",
    "Session4Date",
    "Session4DateUtc",
    "Session5Date",
    "Session5DateUtc",
]

MIN_SUPPORTED_YEAR = 1950
MAX_SUPPORTED_YEAR = datetime.today().year

MIN_SUPPORTED_ROUND = 1  # testing events all share round number = 0
MAX_SUPPORTED_ROUND = 25

MIN_SUPPORTED_SESSION = 1
MAX_SUPPORTED_SESSION = 5

DEFAULT_SESSION = 5  # race

MIN_SUPPORTED_DRIVER_NUMBER = 1
MAX_SUPPORTED_DRIVER_NUMBER = 99

MIN_SUPPORTED_LAP_COUNT = 1
