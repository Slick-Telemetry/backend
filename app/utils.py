# Built-in
import logging
from datetime import datetime

# External
import fastf1


logger = logging.getLogger(__name__)

BACKSTOP = 1954


def get_default_year() -> int:
    """Get the default year for the app.

    NOTE - The default year is defined as the latest year which has data for at least 1 race session.
    """

    default_year = datetime.today().year

    while default_year >= BACKSTOP:
        event_schedule = fastf1.get_event_schedule(default_year, include_testing=False)
        if len(event_schedule) == 0:
            # if no data is found, check previous year
            default_year -= 1
            continue
        try:
            # extract first race's round number.
            # per current (2023-01-22) fastf1/ergast implementation, the value is 1.
            race_event_round_number = event_schedule.iloc[0]["RoundNumber"]
            # create a fastf1.core.Session object
            first_race_event_session_obj = fastf1.get_session(default_year, race_event_round_number, "Race")
            # load session data
            first_race_event_session_obj.load()
        except (ValueError, KeyError) as ex:
            # capture errors occurred during extracting first race's round number and loading the session data
            logger.exception(ex)
            default_year -= 1
        else:
            return default_year
    raise ValueError(f"No data found for years upto {BACKSTOP}")
