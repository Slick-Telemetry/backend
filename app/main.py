import json
from typing import Annotated, Literal

import fastf1
from fastapi import FastAPI, HTTPException, Path, Query, status
from fastapi.middleware.cors import CORSMiddleware
from fastf1.ergast import Ergast

from app.constants import (
    EVENT_SCHEDULE_DATETIME_DTYPE_LIST,
    MAX_SUPPORTED_ROUND,
    MAX_SUPPORTED_SESSION,
    MAX_SUPPORTED_YEAR,
    METADATA_DESCRIPTION,
    MIN_SUPPORTED_ROUND,
    MIN_SUPPORTED_SESSION,
    MIN_SUPPORTED_YEAR,
)
from app.models import HealthCheck, Results, Schedule, Standings
from app.utils import get_default_year

# fastf1.set_log_level("WARNING") # TODO use for production and staging

# Cors Middleware
origins = ["http://localhost:3000"]
# Ergast configuration
ergast = Ergast(result_type="raw", auto_cast=True)


app = FastAPI(
    title="Slick Telemetry API",
    description=METADATA_DESCRIPTION,
    version="0.1.0",
    contact={
        "name": "Slick Telemetry",
        "url": "https://github.com/Slick-Telemetry",
        "email": "",
    },
    license_info={
        "name": "GNU General Public License v3.0",
        "url": "https://www.gnu.org/licenses/gpl-3.0.en.html",
    },
)


app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
    # HTTPSRedirectMiddleware # TODO use for production and staging
)


@app.get(
    "/",
    tags=["root"],
    summary="Read root",
    response_description="Return root info",
    status_code=status.HTTP_200_OK,
)
def read_root():
    return {"we_are": "SlickTelemetry"}


# https://gist.github.com/Jarmos-san/0b655a3f75b698833188922b714562e5
@app.get(
    "/health",
    tags=["healthcheck"],
    summary="Perform a Health Check",
    response_description="Return HTTP Status Code 200 (OK)",
    status_code=status.HTTP_200_OK,
    response_model=HealthCheck,
)
def get_health() -> HealthCheck:
    """
    ## Perform a Health Check
    Endpoint to perform a healthcheck on. This endpoint can primarily be used Docker
    to ensure a robust container orchestration and management is in place. Other
    services which rely on proper functioning of the API service will not deploy if this
    endpoint returns any other HTTP status code except 200 (OK).

    **Returns**:
        HealthCheck: Returns a JSON response with the health status
    """
    return HealthCheck(status="OK")


@app.get(
    "/schedule",
    tags=["schedule"],
    summary="Get events schedule for a Formula 1 calendar year",
    response_description="Return list of events schedule for a Formula 1 calendar year",
    status_code=status.HTTP_200_OK,
    response_model=list[Schedule],
)
def get_schedule(
    year: Annotated[
        int | None,
        Query(
            title="The year for which to get the schedule",
            ge=MIN_SUPPORTED_YEAR,
            le=MAX_SUPPORTED_YEAR,
        ),
    ] = None
) -> list[Schedule]:
    """
    ## Get events schedule for a Formula 1 calendar year
    Endpoint to get events schedule for Formula 1 calendar year.

    **NOTE**: If `year` is not provided; we use the default year. Default year is defined as the year which has data for at least 1 race session.

    **Returns**:
        list[Schedule]: Returns a JSON response with the list of event schedule
    """

    if year is None:
        year = get_default_year()

    event_schedule = fastf1.get_event_schedule(year)

    # Convert timestamp(z) related columns' data into string type
    # https://stackoverflow.com/questions/50404559/python-error-typeerror-object-of-type-timestamp-is-not-json-serializable
    for col in EVENT_SCHEDULE_DATETIME_DTYPE_LIST:
        event_schedule[col] = event_schedule[col].astype(str)

    # Convert the dataframe to a JSON string
    event_schedule_as_json = event_schedule.to_json(orient="records")

    # Parse the JSON string to a JSON object
    event_schedule_as_json_obj = json.loads(event_schedule_as_json)

    return event_schedule_as_json_obj


@app.get(
    "/standings",
    tags=["standings"],
    summary="Get drivers and constructors standings ",
    response_description="Return a list of drivers and constructors standings at specific points of a season. If the season hasn't ended you will get the current standings.",
    status_code=status.HTTP_200_OK,
    response_model=Standings,
)
def get_standings(
    year: Annotated[
        int | None,
        Query(
            title="The year for which to get the driver and constructors standing. If the season hasn't ended you will get the current standings.",
            ge=MIN_SUPPORTED_YEAR,
            le=MAX_SUPPORTED_YEAR,
        ),
    ] = None,
    round: Annotated[
        int | None,
        Query(
            title="The round in a year for which to get the driver and constructor standings",
            ge=MIN_SUPPORTED_ROUND,
            le=MAX_SUPPORTED_ROUND,
        ),
    ] = None,
) -> Standings:
    """
    ## Get driver and constructor standings
    Endpoint to get driver and constructor standings at specific points of a season. If the season hasn't ended you will get the current standings.

    **NOTE**: If `year` is not provided; we use the default year. Default year is defined as the year which has data for at least 1 race session.

    **Returns**:
        Standings: Returns a JSON response with the driver and constructor standings
    """

    if year is None and round is None:
        # neither year nor round are provided; get results for the last round of the default year
        year = get_default_year()
    elif year is None and round is not None:
        # only round is provided; error out
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail='Bad request. Must provide the "year" parameter.',
        )

    # inputs are good; either one of the two remaining cases:
    # 1. both year and round are provided
    # 2. only year is provided

    driver_standings = ergast.get_driver_standings(season=year, round=round)
    constructor_standings = ergast.get_constructor_standings(season=year, round=round)

    driver_standings_available = True if len(driver_standings) > 0 else False
    constructor_standings_available = True if len(constructor_standings) > 0 else False

    if driver_standings_available and constructor_standings_available:
        # both driver and constructor standings are available
        data: Standings = {
            "season": driver_standings[0]["season"],
            "round": driver_standings[0]["round"],
            "DriverStandings": driver_standings[0]["DriverStandings"],
            "ConstructorStandings": constructor_standings[0]["ConstructorStandings"],
        }
        return data
    elif not driver_standings_available and not constructor_standings_available:
        # neither driver nor constructor standings are available
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Driver and constructor standings not found.",
        )
    elif driver_standings_available:
        # only driver standings are available
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Constructor standings not found.",
        )
    elif constructor_standings_available:
        # only constructor standings are available
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Driver standings not found."
        )
    else:
        # something went wrong, investigate
        raise HTTPException(
            status_code=500, detail="Something went wrong. Investigate!"
        )


@app.get(
    "/results/{year}/{round}/{session}",
    tags=["results"],
    summary="Get session results for a given year, round and session",
    response_description="Return session results for a given year, round and session.",
    status_code=status.HTTP_200_OK,
    response_model=list[Results],
)
def get_results(
    year: Annotated[
        int,
        Path(
            title="The year for which to get the results",
            ge=MIN_SUPPORTED_YEAR,
            le=MAX_SUPPORTED_YEAR,
        ),
    ],
    round: Annotated[
        int,
        Path(
            title="The round in a year for which to get the results",
            ge=MIN_SUPPORTED_ROUND,
            le=MAX_SUPPORTED_ROUND,
        ),
    ],
    session: Annotated[
        int,
        Path(
            title="The session in a round for which to get the results",
            ge=MIN_SUPPORTED_SESSION,
            le=MAX_SUPPORTED_SESSION,
        ),
    ],
) -> list[Results]:
    """
    ## Get session results for a given year, round and session
    Endpoint to get session results for a given year, round and session.

    **Returns**:
        list[Results]: Returns a JSON response with the list of session results
    """

    try:
        session_obj = fastf1.get_session(year=year, gp=round, identifier=session)
        session_obj.load(
            laps=True,
            telemetry=False,
            weather=False,
            messages=False,
        )
        session_results = session_obj.results

        # Convert the dataframe to a JSON string
        session_results_as_json = session_results.to_json(orient="records")

        # Parse the JSON string to a JSON object
        session_results_as_json_obj = json.loads(session_results_as_json)
        return session_results_as_json_obj

    except ValueError as ve:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=f"Bad Request. {str(ve)}"
        )
    except KeyError as ke:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=f". {str(ke)}"
        )
