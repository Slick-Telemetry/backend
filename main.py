import json

import fastf1
from fastapi import FastAPI
from fastapi.responses import JSONResponse

from constants import EVENT_SCHEDULE_DATETIME_DTYPE_LIST
from utils import get_default_year

app = FastAPI()


@app.get("/")
def read_root():
    return {"We Are": "SlickTelemetry"}


@app.get("/schedule/{year}")
def schedule(year: int = get_default_year()):
    event_schedule = fastf1.get_event_schedule(year)

    # Convert timestamp(z) related columns' data into a string type
    # https://stackoverflow.com/questions/50404559/python-error-typeerror-object-of-type-timestamp-is-not-json-serializable
    for col in EVENT_SCHEDULE_DATETIME_DTYPE_LIST:
        event_schedule[col] = event_schedule[col].astype(str)

    # Convert the dataframe to a JSON string
    event_schedule_as_json = event_schedule.to_json(orient="records")

    # Parse the JSON string to a JSON object
    event_schedule_as_json_obj = json.loads(event_schedule_as_json)

    return JSONResponse(content=event_schedule_as_json_obj)
