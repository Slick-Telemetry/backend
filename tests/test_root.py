# External
from fastapi.testclient import TestClient

# Project
from app.main import app


client = TestClient(app)


def test_read_root(snapshot):
    response = client.get("/")
    snapshot.assert_match(response.json())
