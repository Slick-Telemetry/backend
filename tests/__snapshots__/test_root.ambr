# serializer version: 1
# name: test_read_root
  dict({
    'we_are': 'SlickTelemetry',
  })
# ---
